crates/package-manager/src/add.rs

```diff
@@ -0,0 +1,88 @@
+use crate::Install;
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::Lockfile;
+use pacquet_npmrc::Npmrc;
+use pacquet_package_manifest::PackageManifestError;
+use pacquet_package_manifest::{DependencyGroup, PackageManifest};
+use pacquet_registry::{PackageTag, PackageVersion};
+use pacquet_tarball::Cache;
+use reqwest::Client;
+
+/// This subroutine does everything `pacquet add` is supposed to do.
+#[must_use]
+pub struct Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    pub tarball_cache: &'a Cache,
+    pub http_client: &'a Client,
+    pub config: &'static Npmrc,
+    pub manifest: &'a mut PackageManifest,
+    pub lockfile: Option<&'a Lockfile>,
+    pub list_dependency_groups: ListDependencyGroups, // must be a function because it is called multiple times
+    pub package_name: &'a str, // TODO: 1. support version range, 2. multiple arguments, 3. name this `packages`
```

Comment:
it is not really a package name only because it may have a version or a tag specified to. Like `express@1`.

in pnpm will call it `package_specifier` or `package_spec`. I think npm calls it a package spec.

and I think Yarn calls it `package_descriptor`.

Reply:
Right now, it is `package_name`. The name would change once we add version range to `pacquet add`, but right now, it only functions as a package name.